aten/src/ATen/native/mobile/cpu/Engine.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+// Prefer not to use these functions directly.  Instead prefer to use the
+// corresponding mobile operators that this module registers with, and exposes
+// through, c10 whenever possible in order to decouple op creation from op
+// execution.  The underlying computation in many NN operators is performed in
```

Comment:
Nooooo these ops have caching?

Reply:
Well that's a considerable part of the performance gain.  If we care about performance, we have to enable the use case where one-time computations are factored out.  The difference is significant (a factor of 2x on MobileNetV2.)

Reply:
By the way, the caching is explicit and immutable, and exposed directly to the user.  It is not an internal global state.

Reply:
I took a look at the "caching" for the add operator and I didn't see any substantive computation, except for error checking: https://github.com/AshkanAliabadi/XNNPACK/blob/1465014503787e1a2f0806a729d511a5c49bb231/src/add.c#L154 and the caching for the convolution operator https://github.com/AshkanAliabadi/XNNPACK/blob/1465014503787e1a2f0806a729d511a5c49bb231/src/convolution.c#L61 which appears to be about prepacking weights. Is there any one-time computation you want to factor out that isn't dependent on tensor inputs?

Reply:
@dzhulgakov Don't we already have a story for prepacking weights in TorchScript?

Reply:
The weight reordering in the convolution operator is a significant portion of the time spent in the convolution operator.  I have added you to an internal document showing a factor of 2x difference on MobileNetV2.

Generally speaking, I would like to point out that I understand and value the motivations behind PyTorch's functional design - both from a code quality and software engineering perspective, and the implications that maintaining internal state has on threading.  Functional code is typically clearer, less error prone, easier to debug, more flexible lending itself to changes, and easier to parallelize.  A functional design that avoids internal state, is a good design.  No arguments there. :)

Having said that, there is a bit of tension between that philosophy and performance at times.  If you subscribe to the idea that good software design, while very important, is secondary to software being useful to the customers, then performance being a critical go / no-go feature of PyTorch Mobile, I think we should agree that redoing the same computation on a performance critical code-path, in an inner loop (while (frame / etc.) { forward(..) }) puts PyTorch at a significant disadvantage compared to the other frameworks out there - mainly because they do not put any such self imposed limitations on their design.

When you are operating under a tight computational budget, as is typically found on mobile and embedded devices, every millisecond matters.  If your goal is to hit 60 frames per second on MobileNetV2 or any other mobile friendly network, then even something as trivial as memory allocations and deallocations, when done on each operator and many times per frame, can easily add up to cost you a millisecond.  As a matter of fact, that is already costing us much more than that as you know, but I am being conservative with that number here.  Regardless of the actual number, that difference means whether this self-driving network, or that AR/VR app can, or cannot run on PyTorch given the accuracy desired.

This is not supposed to be a rant by the way :) - I am just trying to point out that while I value functional design, operating under a tight computation budget does not leave much room for redoing computations.  What we are doing here is to at least make this explicit to the user by breaking the operation into conv_prepack / conv_run operation.

But the idea is even more general than that, I think limiting that to conv_prepack and conv_run is doing us a disservice performance-wise.  You can imagine how many operators can have one-time computations, from one-time memory allocations or creating a look-up table or what have you, that do not need to be re-done on every forward.  Now I understand if we do not want to expose everything at that level to the user because the simplicity-performance tradeoff does not make sense, but at the very least, I think there is value in writing the new operators with that differentiation in mind, so when there is the need to expose it, if ever, we are at least ready.   Here I am supporting both code paths - the separate create() and run(), and the unified simpler create_and_run().

Reply:
You don't have to convince me that we should support weight prepacking. :)

My confusion here is because I haven't been completely following the design arc of weight prepacking. I remember as early as April (e.g., https://fb.quip.com/MU7aAbzGDesu) we had some design documents assessing the various representations for prepacked weights. Now, my recollection was that we agreed that we would support arbitrary structs to represent the packed data (since it usually can't be conveniently represented as just a tensor), and then we would pass these around as first class objects in the IR.

Since then, I haven't seen any diffs which would have added support for this in the eager interpreter (I guess @jamesr66a is working on this JIT side with TorchBind?) Nor have I seen any updated design docs. And since this diff is missing all of the binding bits, it's very hard for me to evaluate if the current organization here actually works or not. Being told ahead of time what the plan is would help a lot!

Reply:
Sorry I didn't include you, that wasn't intentional.  I'll make sure to include you in subsequent design documents / meetings.

Reply:
Just referencing them from the PR is good :)